### PR TITLE
robust rename a file or folder to avoid 'Access is denied' on Windows.

### PR DESCRIPTION
**Pull request by yangcha**
Changelog: Feature: New ``tools.rename`` function to rename a file or folder to avoid 'Access is denied' on Windows.
Docs: https://github.com/conan-io/docs/pull/1646

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conans/client/tools/files.py [hunk 1]
```diff
@@ -383,3 +383,16 @@ def unix2dos(filepath):
...
     _replace_with_separator(filepath, "\n")
+
+def rename(src, dst):
+    """
+    rename a file or folder to avoid "Access is denied" error on Windows
+    :param src: Path to the file or folder
+    :param dst: Destination file or folder
+    """
+    if platform.system() == "Windows":
+        retcode = os.system("robocopy /move /e /ndl /nfl".format(src, dst))
```
Comment (danimtb) [inline conans/client/tools/files.py]: Adding comments about the flags for future review:

```
/E :: copy subdirectories, including Empty ones.
/MOVE :: MOVE files AND dirs (delete from source after copying).
/NFL :: No File List - don't log file names.
/NDL :: No Directory List - don't log directory names.
```

Robocopy is installed by default in Windows even in those without a professional license.

Comment (uilianries) [inline conans/client/tools/files.py]: I would recommend subprocess instead, it's lighter and safer than os.system

Comment (SSE4) [reply conans/client/tools/files.py]: I don't think it's available in all Windows versions. better add some check (`tools.which`)

Comment (danimtb) [reply conans/client/tools/files.py]: I checked that in a home edition version but maybe it is not there in a windows server?

Comment (yangcha) [reply conans/client/tools/files.py]: According to wikipedia: https://en.wikipedia.org/wiki/Robocopy , it is bundled with any version of Windows after XP.

Comment (memsharded) [reply conans/client/tools/files.py]: There are conan users stuck with Windows XP :(

**Review by memsharded — COMMENTED** [review]
One of my concerns about this feature is that we cannot provide a consistent red/failing test first in our test suite that actually proofs that this functionality is required.

I know that systems are complicated, and that ``os.rename`` might fail in some circumstances. What are those circumstances? Why something as core and spread as python ``os.rename()`` would fail that much that requires fixing it on our side? I would prefer to have more understanding of the underlying reasons to add this to Conan codebase.

File: conans/client/tools/files.py (hunk 1 above)
Comment (yangcha) [reply conans/client/tools/files.py]: The latest version of this PR checks for existence of robocopy.

File: conans/client/tools/files.py [hunk 2]
```diff
@@ -363,3 +363,18 @@ def unix2dos(filepath):
...
+def rename(src, dst):
+    """
+    rename a file or folder to avoid "Access is denied" error on Windows
+    :param src: Source file or folder
+    :param dst: Destination file or folder
+    """
+    if platform.system() == "Windows" and which("robocopy") and os.path.isdir(src):
+        if os.path.exists(dst):
+            raise ConanException("rename {} to {} failed, dst exists.".format(src, dst))
+        retcode = os.system("robocopy /move /e /ndl /nfl {} {}".format(src, dst))
```
Comment (jokram) [inline conans/client/tools/files.py]: If source or destination has spaces the command might fail, or?
=> Add quotes for src and dst

Comment (yangcha) [reply conans/client/tools/files.py]: Thanks for pointing out. As @uilianries said, I will replace os.system with subprocess which can handle the spaces in path more gracefully.

Comment (yangcha) [reply conans/client/tools/files.py]: > If source or destination has spaces the command might fail, or?
> => Add quotes for src and dst
Use subprocess and command sequence instead os.system(). I tested and it works for source or destination has spaces.

**Comment by yangcha** [comment]
> One of my concerns about this feature is that we cannot provide a consistent red/failing test first in our test suite that actually proofs that this functionality is required.
> 
> I know that systems are complicated, and that `os.rename` might fail in some circumstances. What are those circumstances? Why something as core and spread as python `os.rename()` would fail that much that requires fixing it on our side? I would prefer to have more understanding of the underlying reasons to add this to Conan codebase.

If you check the source code of os.rename,(), you will see that it recursively copies and deletes sub-folders and files using recursion.   On Windows, this could cause "Access is denied".

File: conans/client/tools/files.py (hunk 1 above)
Comment (yangcha) [reply conans/client/tools/files.py]: > I would recommend subprocess instead, it's lighter and safer than os.system

Switched to subprocess.

File: conans/client/tools/files.py [hunk 3]
```diff
@@ -369,6 +370,23 @@ def dos2unix(filepath):
...
+def rename(src, dst):
+    """
+    rename a file or folder to avoid "Access is denied" error on Windows
+    :param src: Source file or folder
+    :param dst: Destination file or folder
+    """
+    if platform.system() == "Windows" and which("robocopy") and os.path.isdir(src):
+        if os.path.exists(dst):
+            raise ConanException("rename {} to {} failed, dst exists.".format(src, dst))
+        completed = subprocess.run(["robocopy", "/move", "/e", "/ndl", "/nfl", src, dst],
```
Comment (memsharded) [inline conans/client/tools/files.py:387]: Please document in comments the args provided to ``robocopy``

**Review by memsharded — APPROVED** [review]
This looks good, low risk as a new tool (I would probably label it as "experimental" in the docs in the first release, we might learn something tricky about robocopy or something unexpected).

Could you please add some unit test that cover this functionality? Thanks!

File: conans/client/tools/files.py (hunk 2 above)
Comment (memsharded) [reply conans/client/tools/files.py]: Unit tests should cover the spaces thing (create temp folders, or use the TestClient, which automatically adds spaces in the path).

File: conans/client/tools/files.py [hunk 4]
```diff
@@ -369,6 +370,23 @@ def dos2unix(filepath):
...
+                                   stdout=subprocess.PIPE)
+        if completed.returncode != 1:
+            raise ConanException("rename {} to {} failed.".format(src, dst))
+    else:
+        os.rename(src, dst)
```
Comment (jgsogo) [inline conans/client/tools/files.py]: We should return the same type of exception if the `rename` fails. I'm ok if we capture all different errors and raise a `ConanException('rename failed....')`.

**Review by jgsogo — CHANGES_REQUESTED** [review]
Nice to have it as a tool 👍

File: conans/client/tools/files.py (hunk 4 above)
Comment (yangcha) [reply conans/client/tools/files.py]: @jgsogo Added new commit to raise `ConanException` if `os.rename()` fails.

File: conans/client/tools/files.py (hunk 3 above)
Comment (yangcha) [reply conans/client/tools/files.py:387]: Comments on robcopy's parameters are added.

**Comment by yangcha** [comment]
> This looks good, low risk as a new tool (I would probably label it as "experimental" in the docs in the first release, we might learn something tricky about robocopy or something unexpected).
> 
> Could you please add some unit test that cover this functionality? Thanks!

Unit test is added.

File: conans/client/tools/files.py [hunk 5]
```diff
@@ -369,6 +370,33 @@ def dos2unix(filepath):
...
+                                   stdout=subprocess.PIPE)
+        if completed.returncode != 1:
+            raise ConanException("rename {} to {} failed.".format(src, dst))
+    else:
+        try:
+            os.rename(src, dst)
+        except Exception as err:
+            raise ConanException("rename {} to {} failed: {}".format(src, dst, err))
+        except:
+            raise ConanException("rename {} to {} failed, unexpected error.".format(src, dst))
```
Comment (jgsogo) [inline conans/client/tools/files.py]: IMO, this last catch 'em all is not needed. I would expect any error to inherit from `Exception`, have you read something different somewhere for the `os.rename` function?

```suggestion
```

**Review by jgsogo — COMMENTED** [review]
Thanks a lot for adding those comments and unittests.

File: conans/client/tools/files.py (hunk 5 above)
Comment (yangcha) [reply conans/client/tools/files.py]: catch-all exception handling has been removed. No particular reason to add it, just be extremely cautious. :)

**Review by jgsogo — APPROVED** [review]

**Pull request merged**